File: android/src/main/java/health/hsc/plugins/storefront/StorefrontPlugin.java
package health.hsc.plugins.storefront;

import androidx.annotation.NonNull;
import androidx.annotation.VisibleForTesting;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

/** Capacitor bridge for the {@code Storefront} plugin. */
@CapacitorPlugin(name = "Storefront")
public class StorefrontPlugin extends Plugin {

    /** Default timeout in milliseconds. */
    public static final long DEFAULT_TIMEOUT_MS = 10_000L;

    private StorefrontProvider provider;

    @Override
    public void load() {
        if (provider == null) {
            provider = new PlayBillingStorefrontProvider(getContext());
        }
    }

    /** Replaces the Play Billing backed provider, e.g. with a fake in tests. */
    @VisibleForTesting
    public void setStorefrontProvider(@NonNull StorefrontProvider provider) {
        this.provider = provider;
    }

    @PluginMethod
    public void getStorefront(PluginCall call) {
        if (provider == null) {
            provider = new PlayBillingStorefrontProvider(getContext());
        }
        provider.getStorefront(
            timeout(call.getDouble("timeout")),
            new StorefrontProvider.Callback() {
                @Override
                public void onSuccess(@NonNull StorefrontInfo info) {
                    call.resolve(info.toJSObject());
                }

                @Override
                public void onError(@NonNull StorefrontException error) {
                    call.reject(error.getMessage(), error.getCode(), error, error.toData());
                }
            }
        );
    }

    /** Falls back to the default for missing, non-finite or non-positive values. */
    static long timeout(Double option) {
        if (option == null || option.isNaN() || option.isInfinite() || option <= 0) {
            return DEFAULT_TIMEOUT_MS;
        }
        return Math.max(1L, Math.round(option));
    }
}

File: ios/Sources/StorefrontPlugin/StorefrontPlugin.swift
import Foundation
import Capacitor

/// Capacitor bridge for the `Storefront` plugin.
///
/// See https://capacitorjs.com/docs/plugins/ios for the plugin development guide.
@objc(StorefrontPlugin)
public class StorefrontPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "StorefrontPlugin"
    public let jsName = "Storefront"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "getStorefront", returnType: CAPPluginReturnPromise)
    ]
    private let reader = StorefrontReader()

    @objc func getStorefront(_ call: CAPPluginCall) {
        let timeout = Self.timeout(from: call.getDouble("timeout"))
        let reader = self.reader
        Task {
            do {
                let info = try await reader.read(timeout: timeout)
                call.resolve(info.dictionary)
            } catch let error as StorefrontError {
                call.reject(error.message, error.code, error, error.data)
            } catch {
                call.reject(error.localizedDescription, StorefrontError.unavailable(distribution: nil).code, error)
            }
        }
    }

    /// Falls back to the default for missing, non-finite or non-positive values.
    static func timeout(from option: Double?) -> Double {
        guard let option = option, option.isFinite, option > 0 else {
            return StorefrontReader.defaultTimeout
        }
        return option
    }
}
